src/components/InputV2/Address.tsx

```diff
@@ -0,0 +1,82 @@
+import PropTypes from 'prop-types'
+import { useEffect, useState } from 'react'
+import { validateAddressString } from '@glif/filecoin-address'
+import { BaseInput, BaseInputProps, BaseInputPropTypes } from './Base'
+import truncateAddress from '../../utils/truncateAddress'
+
+/**
+ * AddressInput
+ */
+export const AddressInput = ({
+  value,
+  onFocus,
+  onBlur,
+  setIsValid,
+  ...baseProps
+}: AddressInputProps) => {
+  const [error, setError] = useState<string>('')
+  const [hasFocus, setHasFocus] = useState<boolean>(false)
+  const [truncated, setTruncated] = useState<string>('')
+
+  const onFocusBase = () => {
+    setHasFocus(true)
+    onFocus()
+  }
+
+  const onBlurBase = () => {
+    setHasFocus(false)
+    onBlur()
+  }
+
+  useEffect(() => {
+    const isValid = validateAddressString(value)
+    setTruncated(isValid ? truncateAddress(value) : value)
+    setError(isValid ? '' : 'Needs to be a valid address')
+    setIsValid(isValid)
```

Comment:
I'm confused a bit by how this doesn't trigger an infinite loop? I feel like we could:

- remove the stateful `truncated` variable, since this can be computed on the fly (or at least in a `useMemo` 
- Only `setState` inside the `useEffect` when we need to (if isValid status has changed and if an error state should change)

Reply:
I see this pattern is used elsewhere, and wondering if it matters or if react under the hood got smart about this type of thing (setting state to an existing value where the value hasn't changed, doesn't actually trigger a rerender).

https://github.com/glifio/react-components/blob/21d449205ca7c2e9da561d15397d645ed072ae04/src/components/InputV2/Filecoin.tsx#L97-L98 i see it used here too, but missed flagging it

Swear there used to be some linter warning for this

Reply:
This useEffect is only executed when the values in the supplied array change, which we are not setting the state of. So either way I don't think this could cause an infinite loop. But I think you're right that we maybe should use useMemo for the truncated version

Reply:
I've rewritten the validated inputs and removed the local `setState` calls from `useEffect`.

We only still call `setIsValid` from `useEffect` in order to set the `isValid` state on the parent component. I don't see any way around this. I will only cause issues when the parent component modifies `value` based on `isValid`, which could cause the validator to re-run and cause an infinite loop. But I think we can make it a rule that this should never happen, I don't see a situation where that makes sense.